File: fitness-backend/gamification.py
"""게이미피케이션: 포인트 원장과 뱃지.

포인트는 기록에서 매번 다시 계산한다. (사용자, 날짜, 종류)에 유일 제약이 있어
몇 번을 재계산해도 중복 지급되지 않는다. 기록을 지우면 포인트도 따라 줄어든다.
"""

from datetime import timedelta

from sqlalchemy import select

from models import MealLog, PointEvent, SetLog, UserBadge, WorkoutLog

POINTS = {
    "workout_done": 12,
    "workout_partial": 6,
    "meal_per_log": 3,
    "meal_daily_cap": 12,
    "protein_hit": 8,
}

LEVELS = [
    (0, "입문자"),
    (150, "습관 형성"),
    (400, "꾸준러"),
    (800, "중급 트레이니"),
    (1500, "베테랑"),
    (2500, "철인"),
]

# 뱃지 카탈로그. condition은 stats dict를 받아 (달성 여부, 현재값, 목표값)을 돌려준다.
BADGES = [
    {
        "code": "first_step",
        "name": "첫 걸음",
        "icon": "🏁",
        "description": "첫 운동 기록을 남겼습니다.",
        "metric": lambda s: (s["workout_days"], 1),
    },
    {
        "code": "streak_3",
        "name": "3일 연속",
        "icon": "🔥",
        "description": "3일 연속으로 기록했습니다.",
        "metric": lambda s: (s["best_streak"], 3),
    },
    {
        "code": "streak_7",
        "name": "일주일 완주",
        "icon": "🔥",
        "description": "7일 연속으로 기록했습니다.",
        "metric": lambda s: (s["best_streak"], 7),
    },
    {
        "code": "volume_10k",
        "name": "1만 kg",
        "icon": "🏋️",
        "description": "누적 볼륨 10,000kg을 들어올렸습니다.",
        "metric": lambda s: (s["total_volume"], 10000),
    },
    {
        "code": "volume_50k",
        "name": "5만 kg",
        "icon": "🏔️",
        "description": "누적 볼륨 50,000kg을 넘겼습니다.",
        "metric": lambda s: (s["total_volume"], 50000),
    },
    {
        "code": "protein_5",
        "name": "단백질 수호자",
        "icon": "🥩",
        "description": "단백질 목표를 90% 이상 채운 날이 5일입니다.",
        "metric": lambda s: (s["protein_days"], 5),
    },
    {
        "code": "meal_30",
        "name": "기록 습관",
        "icon": "🍽️",
        "description": "식단을 30끼 기록했습니다.",
        "metric": lambda s: (s["meal_count"], 30),
    },
    {
        "code": "comeback",
        "name": "다시 시작",
        "icon": "💪",
        "description": "빠진 다음 날 바로 복귀했습니다.",
        "metric": lambda s: (s["comebacks"], 1),
    },
    {
        "code": "honest_logger",
        "name": "정직한 기록자",
        "icon": "📝",
        "description": "체감 강도(RPE)를 5회 이상 남겼습니다.",
        "metric": lambda s: (s["rpe_logs"], 5),
    },
]


def _level_for(points):
    title = LEVELS[0][1]
    level = 1
    next_at = LEVELS[1][0]

    for index, (threshold, name) in enumerate(LEVELS):
        if points >= threshold:
            level = index + 1
            title = name
            next_at = LEVELS[index + 1][0] if index + 1 < len(LEVELS) else None

    return {"level": level, "title": title, "nextAt": next_at}


def _streaks(dates):
    """연속 기록일: (현재 연속, 최고 연속)."""
    if not dates:
        return 0, 0

    ordered = sorted(dates)
    best = 1
    run = 1
    for prev, current in zip(ordered, ordered[1:]):
        if (current - prev).days == 1:
            run += 1
            best = max(best, run)
        else:
            run = 1

    # 마지막 구간이 현재 연속이다.
    return run, best


def collect_stats(db, user_id, protein_target):
    workouts = db.scalars(
        select(WorkoutLog).where(WorkoutLog.user_id == user_id).order_by(WorkoutLog.log_date)
    ).all()
    meals = db.scalars(select(MealLog).where(MealLog.user_id == user_id)).all()

    workout_ids = [log.id for log in workouts]
    sets = (
        db.scalars(select(SetLog).where(SetLog.workout_log_id.in_(workout_ids))).all()
        if workout_ids
        else []
    )

    total_volume = sum(entry.weight_kg * entry.reps for entry in sets)

    protein_by_date = {}
    for meal in meals:
        protein_by_date[meal.log_date] = protein_by_date.get(meal.log_date, 0) + meal.protein
    protein_days = sum(
        1 for value in protein_by_date.values() if protein_target and value >= protein_target * 0.9
    )

    log_dates = {log.log_date for log in workouts if log.status != "missed"}
    log_dates |= set(protein_by_date.keys())
    current_streak, best_streak = _streaks(list(log_dates))

    # 미수행 다음 날 완료 = 복귀
    by_date = {log.log_date: log.status for log in workouts}
    comebacks = sum(
        1
        for log_date, status in by_date.items()
        if status == "missed" and by_date.get(log_date + timedelta(days=1)) == "done"
    )

    return {
        "workouts": workouts,
        "meals": meals,
        "workout_days": sum(1 for log in workouts if log.status in ("done", "partial")),
        "done_days": sum(1 for log in workouts if log.status == "done"),
        "total_volume": round(total_volume),
        "meal_count": len(meals),
        "protein_days": protein_days,
        "protein_by_date": protein_by_date,
        "current_streak": current_streak,
        "best_streak": best_streak,
        "comebacks": comebacks,
        "rpe_logs": sum(1 for log in workouts if log.rpe),
    }


def sync_points(db, user_id, stats, protein_target):
    """기록에서 포인트 원장을 다시 만든다. 이미 있는 항목은 금액만 맞춘다."""
    existing = {
        (row.log_date, row.kind): row
        for row in db.scalars(select(PointEvent).where(PointEvent.user_id == user_id)).all()
    }
    wanted = {}

    for log in stats["workouts"]:
        if log.status == "done":
            wanted[(log.log_date, "workout")] = (POINTS["workout_done"], "운동 완료")
        elif log.status == "partial":
            wanted[(log.log_date, "workout")] = (POINTS["workout_partial"], "운동 부분 수행")

    meals_by_date = {}
    for meal in stats["meals"]:
        meals_by_date[meal.log_date] = meals_by_date.get(meal.log_date, 0) + 1

    for log_date, count in meals_by_date.items():
        amount = min(count * POINTS["meal_per_log"], POINTS["meal_daily_cap"])
        wanted[(log_date, "meal")] = (amount, f"식단 {count}끼 기록")

    if protein_target:
        for log_date, protein in stats["protein_by_date"].items():
            if protein >= protein_target * 0.9:
                wanted[(log_date, "protein")] = (POINTS["protein_hit"], "단백질 목표 달성")

    for key, (amount, reason) in wanted.items():
        row = existing.get(key)
        if row is None:
            db.add(
                PointEvent(
                    user_id=user_id,
                    log_date=key[0],
                    kind=key[1],
                    amount=amount,
                    reason=reason,
                )
            )
        elif row.amount != amount:
            row.amount = amount
            row.reason = reason

    # 기록이 사라진 항목은 원장에서도 지운다.
    for key, row in existing.items():
        if key not in wanted:
            db.delete(row)

    db.commit()

    return sum(amount for amount, _ in wanted.values())


def sync_badges(db, user_id, stats):
    owned = {
        row.code: row
        for row in db.scalars(select(UserBadge).where(UserBadge.user_id == user_id)).all()
    }

    result = []
    for badge in BADGES:
        current, target = badge["metric"](stats)
        earned = current >= target

        if earned and badge["code"] not in owned:
            row = UserBadge(user_id=user_id, code=badge["code"])
            db.add(row)
            owned[badge["code"]] = row

        result.append({
            "code": badge["code"],
            "name": badge["name"],
            "icon": badge["icon"],
            "description": badge["description"],
            "earned": badge["code"] in owned or earned,
            "progress": {"current": min(current, target), "target": target},
        })

    db.commit()

    for item in result:
        row = owned.get(item["code"])
        item["earnedAt"] = row.earned_at.isoformat() if row and row.earned_at else None

    return result


def build_summary(db, user_id, protein_target):
    stats = collect_stats(db, user_id, protein_target)
    points = sync_points(db, user_id, stats, protein_target)
    badges = sync_badges(db, user_id, stats)

    recent = db.scalars(
        select(PointEvent)
        .where(PointEvent.user_id == user_id)
        .order_by(PointEvent.log_date.desc(), PointEvent.id.desc())
        .limit(8)
    ).all()

    return {
        "points": points,
        **_level_for(points),
        "currentStreak": stats["current_streak"],
        "bestStreak": stats["best_streak"],
        "totalVolume": stats["total_volume"],
        "workoutDays": stats["workout_days"],
        "mealCount": stats["meal_count"],
        "badges": badges,
        "earnedCount": sum(1 for badge in badges if badge["earned"]),
        "recent": [
            {
                "date": row.log_date.isoformat(),
                "kind": row.kind,
                "amount": row.amount,
                "reason": row.reason,
            }
            for row in recent
        ],
    }

File: fitness-backend-js/src/domain/gamification.js
/**
 * 게이미피케이션: 포인트 원장과 뱃지.
 *
 * 포인트는 기록에서 매번 다시 계산한다. (사용자, 날짜, 종류)에 유일 제약이 있어
 * 몇 번을 재계산해도 중복 지급되지 않는다.
 */

import { and, eq, inArray } from "drizzle-orm";

import { db } from "../core/db.js";
import {
  mealLogs,
  pointEvents,
  setLogs,
  userBadges,
  workoutLogs,
} from "../db/schema.js";
import { isoOut, nowIso } from "../core/num.js";
import { LEVELS, POINTS } from "./data/gamification.data.js";

export { POINTS, LEVELS };

/** 뱃지 카탈로그. metric은 stats를 받아 [현재값, 목표값]을 돌려준다. */
export const BADGES = [
  {
    code: "first_step",
    name: "첫 걸음",
    icon: "🏁",
    description: "첫 운동 기록을 남겼습니다.",
    metric: (s) => [s.workoutDays, 1],
  },
  {
    code: "streak_3",
    name: "3일 연속",
    icon: "🔥",
    description: "3일 연속으로 기록했습니다.",
    metric: (s) => [s.bestStreak, 3],
  },
  {
    code: "streak_7",
    name: "일주일 완주",
    icon: "🔥",
    description: "7일 연속으로 기록했습니다.",
    metric: (s) => [s.bestStreak, 7],
  },
  {
    code: "volume_10k",
    name: "1만 kg",
    icon: "🏋️",
    description: "누적 볼륨 10,000kg을 들어올렸습니다.",
    metric: (s) => [s.totalVolume, 10000],
  },
  {
    code: "volume_50k",
    name: "5만 kg",
    icon: "🏔️",
    description: "누적 볼륨 50,000kg을 넘겼습니다.",
    metric: (s) => [s.totalVolume, 50000],
  },
  {
    code: "protein_5",
    name: "단백질 수호자",
    icon: "🥩",
    description: "단백질 목표를 90% 이상 채운 날이 5일입니다.",
    metric: (s) => [s.proteinDays, 5],
  },
  {
    code: "meal_30",
    name: "기록 습관",
    icon: "🍽️",
    description: "식단을 30끼 기록했습니다.",
    metric: (s) => [s.mealCount, 30],
  },
  {
    code: "comeback",
    name: "다시 시작",
    icon: "💪",
    description: "빠진 다음 날 바로 복귀했습니다.",
    metric: (s) => [s.comebacks, 1],
  },
  {
    code: "honest_logger",
    name: "정직한 기록자",
    icon: "📝",
    description: "체감 강도(RPE)를 5회 이상 남겼습니다.",
    metric: (s) => [s.rpeLogs, 5],
  },
];

function levelFor(points) {
  let title = LEVELS[0][1];
  let level = 1;
  let nextAt = LEVELS[1][0];

  LEVELS.forEach(([threshold, name], index) => {
    if (points >= threshold) {
      level = index + 1;
      title = name;
      nextAt = index + 1 < LEVELS.length ? LEVELS[index + 1][0] : null;
    }
  });

  return { level, title, nextAt };
}

function dayNumber(iso) {
  return Math.round(new Date(`${iso}T00:00:00Z`).getTime() / 86400000);
}

function shiftDate(iso, days) {
  const date = new Date(`${iso}T00:00:00Z`);
  date.setUTCDate(date.getUTCDate() + days);
  return date.toISOString().slice(0, 10);
}

/** 연속 기록일: [현재 연속, 최고 연속] */
function streaks(dates) {
  if (dates.length === 0) return [0, 0];

  const ordered = [...dates].sort();
  let best = 1;
  let run = 1;

  for (let i = 1; i < ordered.length; i += 1) {
    if (dayNumber(ordered[i]) - dayNumber(ordered[i - 1]) === 1) {
      run += 1;
      best = Math.max(best, run);
    } else {
      run = 1;
    }
  }

  // 마지막 구간이 현재 연속이다.
  return [run, best];
}

export function collectStats(userId, proteinTarget) {
  const workouts = db
    .select()
    .from(workoutLogs)
    .where(eq(workoutLogs.userId, userId))
    .orderBy(workoutLogs.logDate)
    .all();

  const meals = db.select().from(mealLogs).where(eq(mealLogs.userId, userId)).all();

  const workoutIds = workouts.map((log) => log.id);
  const sets = workoutIds.length
    ? db.select().from(setLogs).where(inArray(setLogs.workoutLogId, workoutIds)).all()
    : [];

  const totalVolume = sets.reduce((sum, e) => sum + e.weightKg * e.reps, 0);

  const proteinByDate = new Map();
  for (const meal of meals) {
    proteinByDate.set(
      meal.logDate,
      (proteinByDate.get(meal.logDate) ?? 0) + meal.protein,
    );
  }

  const proteinDays = [...proteinByDate.values()].filter(
    (value) => proteinTarget && value >= proteinTarget * 0.9,
  ).length;

  const logDates = new Set([
    ...workouts.filter((log) => log.status !== "missed").map((log) => log.logDate),
    ...proteinByDate.keys(),
  ]);
  const [currentStreak, bestStreak] = streaks([...logDates]);

  // 미수행 다음 날 완료 = 복귀
  const byDate = new Map(workouts.map((log) => [log.logDate, log.status]));
  let comebacks = 0;
  for (const [logDate, status] of byDate) {
    if (status === "missed" && byDate.get(shiftDate(logDate, 1)) === "done") comebacks += 1;
  }

  return {
    workouts,
    meals,
    workoutDays: workouts.filter((log) => ["done", "partial"].includes(log.status)).length,
    doneDays: workouts.filter((log) => log.status === "done").length,
    totalVolume: Math.round(totalVolume),
    mealCount: meals.length,
    proteinDays,
    proteinByDate,
    currentStreak,
    bestStreak,
    comebacks,
    rpeLogs: workouts.filter((log) => log.rpe).length,
  };
}

/** 기록에서 포인트 원장을 다시 만든다. 이미 있는 항목은 금액만 맞춘다. */
export function syncPoints(userId, stats, proteinTarget) {
  const existing = new Map(
    db
      .select()
      .from(pointEvents)
      .where(eq(pointEvents.userId, userId))
      .all()
      .map((row) => [`${row.logDate}|${row.kind}`, row]),
  );

  const wanted = new Map();

  for (const log of stats.workouts) {
    if (log.status === "done") {
      wanted.set(`${log.logDate}|workout`, [POINTS.workout_done, "운동 완료"]);
    } else if (log.status === "partial") {
      wanted.set(`${log.logDate}|workout`, [POINTS.workout_partial, "운동 부분 수행"]);
    }
  }

  const mealsByDate = new Map();
  for (const meal of stats.meals) {
    mealsByDate.set(meal.logDate, (mealsByDate.get(meal.logDate) ?? 0) + 1);
  }
  for (const [logDate, count] of mealsByDate) {
    const amount = Math.min(count * POINTS.meal_per_log, POINTS.meal_daily_cap);
    wanted.set(`${logDate}|meal`, [amount, `식단 ${count}끼 기록`]);
  }

  if (proteinTarget) {
    for (const [logDate, protein] of stats.proteinByDate) {
      if (protein >= proteinTarget * 0.9) {
        wanted.set(`${logDate}|protein`, [POINTS.protein_hit, "단백질 목표 달성"]);
      }
    }
  }

  for (const [key, [amount, reason]] of wanted) {
    const [logDate, kind] = key.split("|");
    const row = existing.get(key);

    if (!row) {
      db.insert(pointEvents).values({ userId, logDate, kind, amount, reason, createdAt: nowIso() }).run();
    } else if (row.amount !== amount) {
      db.update(pointEvents).set({ amount, reason }).where(eq(pointEvents.id, row.id)).run();
    }
  }

  // 기록이 사라진 항목은 원장에서도 지운다.
  for (const [key, row] of existing) {
    if (!wanted.has(key)) db.delete(pointEvents).where(eq(pointEvents.id, row.id)).run();
  }

  return [...wanted.values()].reduce((sum, [amount]) => sum + amount, 0);
}

export function syncBadges(userId, stats) {
  const owned = new Map(
    db
      .select()
      .from(userBadges)
      .where(eq(userBadges.userId, userId))
      .all()
      .map((row) => [row.code, row]),
  );

  const result = [];
  for (const badge of BADGES) {
    const [current, target] = badge.metric(stats);
    const earned = current >= target;

    if (earned && !owned.has(badge.code)) {
      const row = db
        .insert(userBadges)
        .values({ userId, code: badge.code, earnedAt: nowIso() })
        .returning()
        .get();
      owned.set(badge.code, row);
    }

    result.push({
      code: badge.code,
      name: badge.name,
      icon: badge.icon,
      description: badge.description,
      earned: owned.has(badge.code) || earned,
      progress: { current: Math.min(current, target), target },
      earnedAt: isoOut(owned.get(badge.code)?.earnedAt),
    });
  }

  return result;
}

export function buildSummary(userId, proteinTarget) {
  const stats = collectStats(userId, proteinTarget);
  const points = syncPoints(userId, stats, proteinTarget);
  const badges = syncBadges(userId, stats);

  const recent = db
    .select()
    .from(pointEvents)
    .where(eq(pointEvents.userId, userId))
    .all()
    .sort((a, b) => (a.logDate < b.logDate ? 1 : a.logDate > b.logDate ? -1 : b.id - a.id))
    .slice(0, 8);

  return {
    points,
    ...levelFor(points),
    currentStreak: stats.currentStreak,
    bestStreak: stats.bestStreak,
    totalVolume: stats.totalVolume,
    workoutDays: stats.workoutDays,
    mealCount: stats.mealCount,
    badges,
    earnedCount: badges.filter((badge) => badge.earned).length,
    recent: recent.map((row) => ({
      date: row.logDate,
      kind: row.kind,
      amount: row.amount,
      reason: row.reason,
    })),
  };
}

export { and };
